src/wasm/wasm-interpreter.cpp

```diff
@@ -19,4 +35,252 @@ void Indenter::print() {
+  // when replacing as long as the function does not have any side effects.
+  // Might yield something useful for simple functions like `clamp`, sometimes
+  // even if arguments are only partially constant or not constant at all.
+
+  // Only traverse into functions if not in a function-parallel scenario, where
+  // functions may or may not have been optimized already to something we can
+  // traverse successfully.
+  if (!(flags & FlagValues::PARALLEL)) {
+    // Note that we are not a validator, so we skip calls to functions that do
+    // not exist yet or where the signature does not match.
+    auto* func = module->getFunctionOrNull(curr->target);
+    if (func != nullptr && !func->imported()) {
+      if (func->sig.results.isConcrete()) {
+        auto numOperands = curr->operands.size();
+        if (numOperands == func->getNumParams()) {
+          ContextAwareExpressionRunner runner(module, flags, maxDepth);
+          runner.depth = depth + 1;
+          for (Index i = 0; i < numOperands; ++i) {
+            auto argFlow = visit(curr->operands[i]);
+            if (!argFlow.breaking() && argFlow.values.isConcrete()) {
+              runner.localValues[i] = std::move(argFlow.values);
+            }
+          }
+          auto retFlow = runner.visit(func->body);
+          if (retFlow.breakTo == RETURN_FLOW) {
+            return Flow(std::move(retFlow.values));
+          } else if (!retFlow.breaking()) {
+            return retFlow;
+          }
+        }
+      }
+    }
+  }
+  return Flow(NONCONSTANT_FLOW);
+}
+
+Flow ContextAwareExpressionRunner::visitCallIndirect(CallIndirect* curr) {
+  return Flow(NONCONSTANT_FLOW);
+}
+
+Flow ContextAwareExpressionRunner::visitLocalGet(LocalGet* curr) {
+  // Check if a constant value has been set in the context of this runner.
+  auto iter = localValues.find(curr->index);
+  if (iter != localValues.end()) {
+    return Flow(std::move(iter->second));
+  }
+  // If not the case, see if the calling pass did compute the value of this
+  // specific `local.get` in an earlier step already. This is a fallback
+  // targeting the precompute pass specifically, which already did this work,
+  // but is not applicable when the runner is used via the C-API for example.
+  if (getValues != nullptr) {
+    auto iter = getValues->find(curr);
+    if (iter != getValues->end()) {
+      auto values = iter->second;
+      if (values.isConcrete()) {
+        return Flow(std::move(values));
+      }
+    }
+  }
+  return Flow(NONCONSTANT_FLOW);
+}
+
+Flow ContextAwareExpressionRunner::visitLocalSet(LocalSet* curr) {
+  if (!(flags & FlagValues::REPLACE)) {
+    // If we are evaluating and not replacing the expression, see if there is
+    // a value flowing through a tee.
+    if (curr->type.isConcrete()) {
+      assert(curr->isTee());
+      return visit(curr->value);
+    }
+    // Otherwise remember the constant value set, if any, for subsequent gets.
```

Comment:
Not sure I understand the model here. This seems to assume there is no control flow at all, i.e., nothing like
```
if (..) {
  x = 10;
}
f(x); // x might or might not be 10
```
Is that the assumption? How are we sure it's enforced?

In particular, the name of the class doesn't imply this only considers linear execution traces. If it were `LinearExpressionRunner` then this would be unsurprising.

Perhaps we do want separate enough things for a linear execution runner that it could derive from the context aware one?

Reply:
This got me thinking. Iiuc the base `ExpressionRunner` is already linear, only traversing branches it knows are taken. Did I miss something there? If not, an alternative strategy might be to move all of the new runner into the base `ExpressionRunner` as its default implementation, with deriving runners overriding what they need to override, like one for precompute that uses `getValues`?

Reply:
Oh yes, after thinking on this, I think you're right. We should only be traversing code linearly, and halting if we hit anything that isn't linear anyhow. So yes, we should be able to always track locals that are set and read.

Ok, then yes, I think we can move that into the main runner, which will make this reasonably simple I hope.

Reply:
(and sorry for the delay here!)